File: SIONlib/exercise_2.c
#include <stdlib.h>
#include <mpi.h>
#include <sion.h>

#define ARRAY_SIZE 1000

int main(int argc, char** argv) {
  int * data;
  int i;

  /* SION parameters */
  int        sid, numFiles, globalrank;
  sion_int64 chunksize;
  sion_int32 fsblksize;
  MPI_Comm   lComm = MPI_COMM_NULL;

  /* initialize MPI */
  MPI_Init(&argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &globalrank);

  /* create data */
  data = malloc(ARRAY_SIZE * sizeof(int));
  for(i=0; i<ARRAY_SIZE; ++i) {
     data[i] = globalrank;
  }

  /* create a new file */
  chunksize  = ARRAY_SIZE * sizeof(int);
  numFiles  = 1;
  fsblksize = -1;
  sid = sion_paropen_mpi("parfile.sion", "w", &numFiles,
                         MPI_COMM_WORLD, &lComm,
                         &chunksize, &fsblksize, &globalrank,
                         NULL, NULL);

  /* write data to file */
  sion_fwrite(/*TODO*/);

  /* close file */
  sion_parclose_mpi(sid);

  free(data);

  /* finalize MPI */
  MPI_Finalize();

  return EXIT_SUCCESS;
}

File: SIONlib/solutions/exercise_2.f90
program exercise_2
    use mpi
    use sion_f90
    use sion_f90_mpi

    implicit none
    integer(kind=8) :: chunksize, bwrote
    integer :: sid,globalrank,ierr
    integer :: lComm = MPI_COMM_NULL
    integer(kind=4) :: fsblksize
    integer :: nfiles
    character(len=255) :: newfname

    integer, parameter :: ARRAY_SIZE = 1000
    integer, dimension(:), allocatable :: data

    ! MPI inititalization
    call MPI_Init(ierr)
    call MPI_Comm_rank(MPI_COMM_WORLD,globalrank,ierr)

    ! create data
    allocate(data(ARRAY_SIZE))
    data = globalrank

    ! create a new file
    fsblksize=-1
    chunksize=ARRAY_SIZE*kind(data(1))
    nfiles=1
    call fsion_paropen_mpi("parfile.sion","w",nfiles, MPI_COMM_WORLD, &
                           lComm,chunksize,fsblksize,globalrank,&
                           newfname,sid)

    call fsion_write(data(1),int(kind(data(1)),8),int(ARRAY_SIZE,8),sid,bwrote)

    ! close file 
    call fsion_parclose_mpi(sid,ierr)

    deallocate(data)

    ! MPI finalization
    call MPI_Finalize(ierr)
end
